llama.cpp

```diff
@@ -7783,6 +7783,72 @@ void llama_sample_typical(struct llama_context * ctx, llama_token_data_array * c
     }
 }
 
+void llama_sample_entropy(struct llama_context * ctx, llama_token_data_array * candidates_p, float min_temp, float max_temp, float exponent_val) {
+    const int64_t t_start_sample_us = ggml_time_us();
+
+    llama_sample_softmax(ctx, candidates_p);
+
+    // Calculate entropy of the softmax probabilities
+    float entropy = 0.0f;
+    for (size_t i = 0; i < candidates_p->size; ++i) {
+        float prob = candidates_p->data[i].p;
+        if (prob > 0.0f) { // Ensure no log(0)
+            entropy -= prob * logf(prob);
+        }
+    }
+
+    // Calculate maximum possible entropy
+    float max_entropy = -logf(1.0f / candidates_p->size);
+
+    // Guard against division by zero
+    if (max_entropy == 0.0f) {
+        max_entropy = 1.0f; // This ensures that normalized_entropy will be 0 when entropy is 0
+    }
+
+    // Normalize the entropy
+    float normalized_entropy = entropy / max_entropy;
+
+    // Map the normalized entropy to the desired temperature range using the power function
+    float dyn_temp = min_temp + (max_temp - min_temp) * powf(normalized_entropy, exponent_val);
+
+    // //todo: Ensure to hide print statements unless debugging!
+    // printf("Your text maxtemp value is: %f\n", max_temp);
+    // // Print the variables
+    // printf("Entropy: %f\n", entropy);
+    // printf("Max Possible Entropy: %f\n", max_entropy);
+    // printf("Normalized Entropy: %f\n", normalized_entropy);
+    // printf("Exponent: %f\n", exponent_val);
+    // printf("Dynamic Temperature (dyn_temp): %f\n", dyn_temp);
```

Comment:
Could you please do something about these commented out debug printfs? Either guard them with an #ifdef, or remove them.

Reply:
I guarded them with #ifdef DEBUG